### PR TITLE
buildRustPackage: support custom targets

### PR DESCRIPTION
**Pull request by aaronjanse**
Replaces https://github.com/NixOS/nixpkgs/pull/95405. Ping @Ericson2314 :-)

I'd like feedback for how to refactor the changes in this PR. I do not recommend this PR is merged as-is. Possible changes might be:
- using linkFarm instead of heredocs
- sysroot code could be in its own file

###### Motivation for this change

tl;dr this makes it possible to write Nix derivations to compile many bare-metal Rust projects

I'm working to compile the Redox kernel using Nix, which involves using a custom json target ([x86_64-unknown-none.json](https://gitlab.redox-os.org/redox-os/kernel/-/blob/2057b889ae10bfc9dc0c2bff79e58bcfa4f6c575/targets/x86_64-unknown-none.json)) in buildRustPackage.

At the end of the compile process, the target json path is not "shortened" ([relevant Cargo code](https://github.com/rust-lang/cargo/blob/964a16a28e234a3d397b2a7031d4ab4a428b1391/src/cargo/core/compiler/compile_kind.rs#L151-L168)), causing a failure:
```
    Finished dev [optimized + debuginfo] target(s) in 42.60s
cp: missing destination file operand after 'target//nix/store/dk55xqrgyavw1gizs9b07p71j7przki0-x86_64-unknown-none.json/debug-tmp/'
Try 'cp --help' for more information.
builder for '/nix/store/dhpb8p1nibwr8hw95rymlipcqvw3swrs-redox-iso-latest-x86_64-unknown-redox.drv' failed with exit code 1
error: build of '/nix/store/dhpb8p1nibwr8hw95rymlipcqvw3swrs-redox-iso-latest-x86_64-unknown-redox.drv' failed
```

Also, compiling to a custom Rust target requires re-compiling the Rust standard library, the code for which is included in this PR.

###### Things done

- added handling for when a buildRustPackage target is defined by a JSON file
- added sysroot generation for custom targets
- avoided causing existing Rust packages to need to be recompiled

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by aaronjanse** [comment]
P.S. @Ericson2314 Redox OS can now be declaratively compiled using Nix! (warning: very long build time)

```bash
$ nix-build https://github.com/nix-community/redoxpkgs/archive/vmdisk.tar.gz -A pkgsCross.x86_64-unknown-redox.redox-vmdisk
$ cp result/harddisk.bin .
$ chmod +rw harddisk.bin
$ qemu-system-x86_64 -serial mon:stdio -d cpu_reset -d guest_errors -smp 4 -m 1024 -s -machine q35 -device ich9-intel-hda -device hda-duplex -net nic,model=e1000 -net user -device nec-usb-xhci,id=xhci -device usb-tablet,bus=xhci.0 -enable-kvm -cpu host -drive file=harddrive.bin,format=raw
```

The magic is can be found at https://github.com/nix-community/redoxpkgs/tree/vmdisk/pkgs/redox-vmdisk

**Comment by Ericson2314** [comment]
Sorry for not seeing this earlier! (Feel free to harass me on IRC, your work is very important to me :)).

**Comment by Ericson2314** [comment]
So this is great, first of all. Really excited.

The specific thing I was talking about is a bit different, but I think will readily combine with this. I startied writing some code to explain what I mean:
```diff
diff --git a/pkgs/development/compilers/rust/default.nix b/pkgs/development/compilers/rust/default.nix
index d08b63dd643..32102b03be4 100644
--- a/pkgs/development/compilers/rust/default.nix
+++ b/pkgs/development/compilers/rust/default.nix
@@ -13,14 +13,36 @@
 , llvmPackages_5
 , pkgsBuildTarget, pkgsBuildBuild
 }: rec {
-  toRustTarget = platform: with platform.parsed; let
-    cpu_ = {
-      "armv7a" = "armv7";
-      "armv7l" = "armv7";
-      "armv6l" = "arm";
-    }.${cpu.name} or platform.rustc.arch or cpu.name;
-  in platform.rustc.config
-    or "${cpu_}-${vendor.name}-${kernel.name}${lib.optionalString (abi.name != "unknown") "-${abi.name}"}";
+  _rustTargetJson = platform:
+    builtins.removeAttrs (platform.rustc or {}) [ "name" ];
+
+  _toRustTargetJsonOpt = platform: let
+      targetJson = _rustTargetJson platform;
+      targetFileJsonContents = builtins.toJSON targetJson;
+      targetFileName = (platform.rustc.name or (builtins.hash "sha256" targetFileJsonContents)) + ".json";
+    in
+      if targetJson == {}
+      then null
+      else [ targetFileName targetFileJsonContents ];
+
+  toRustTarget = platform: let
+    noJsonDefaultTargetName = with platform.parsed; let
+      cpu_ = {
+        "armv7a" = "armv7";
+        "armv7l" = "armv7";
+        "armv6l" = "arm";
+      }.${cpu.name} or platform.rustc.arch or cpu.name;
+    in platform.rustc.config
+      or "${cpu_}-${vendor.name}-${kernel.name}${lib.optionalString (abi.name != "unknown") "-${abi.name}"}";
+    jsonAndNameOpt = _toRustTargetJsonOpt platform;
+  in
+    if jsonAndNameOpt == null
+    then noJsonDefaultTargetName
+    else builtins.head jsonAndNameOpt;
+
+  # Don't builtins.readFile this, else eval depends on derivation building
+  # which sadly is not allowed for nixpkgs.
+  toRustTargetTomlFileOpt = platform: ... (_toRustTargetJsonOpt platform) ...;
 
   makeRustPlatform = { rustc, cargo, ... }: rec {
     rust = {
```
So fill in `toRustTargetTomlFileOpt` with what you've got, following the comment, and then use this back in the other file.

Hopefully with https://github.com/NixOS/nix/issues/3929 we can stop worrying about `builtins.readFile` and get rid of the heredoc, but I'm afraid that what you've got so far is largely unavoidable.

I suppose there is one alternative: keep everything in nix until you do one big `toToml` for *everything*. You can try that if you like, but I don't much care either way.

**Comment by Ericson2314** [comment]
CC @cleverca22 & @mic92. Looks like we are on the cusp of proper custom target cross-compilation support!

**Comment by aaronjanse** [comment]
Hmm, I'm having trouble understanding that diff. Are you suggesting that the the target JSON be stored in `stdenv.hostPlatform.rustc`? I'd also like to note that the TOML code is only used for sysroot code (needed for building targets that rustc itself wasn't compiled for). TOML isn't used for the target JSON.

Either way, based on crossing paths with you online, you seem to be much more experienced with this.

**Comment by Ericson2314** [comment]
>  Are you suggesting that the target JSON be stored in stdenv.hostPlatform.rustc?

Yup!

> I'd also like to note that the TOML code is only used for sysroot code (needed for building targets that rustc itself wasn't compiled for).

Ah, I didn't realize that. So it is TOML for sysroot, and JSON for everything else?

> Either way, based on crossing paths with the work you've done with Rust for custom targets, you seem to be much more experienced with this than I.

I'm really rusty with the details of target JSON etc, vs general idea / nixpkgs-specific stuff, so don't take me too seriously on the nitty-gritty! The basic ideas are:

 - Let's try to specify all target-config-like stuff from `stdenv.*Platform`. `platform.rustc` is directly inspired by `platform.gcc`.
   - Always better to have structured values than files and strings
   - Because `build.readFile ...derivation...` and `import ...derivation...` are prohibited for stuff Nixpkgs' hydra builds, let's try to avoid "baking" config to files until as late as possible.
 - Sysrooots are kinda evil, but definitely worth it for now until https://github.com/rust-lang/wg-cargo-std-aware wraps up.
     - Over time, with the work of that working group, standard library and even rustc itself should be built with `buildRustPackage` or `buildRustCrate` (or whatever the latest and great riff on those is).

You're doing great so far :), moving the config to `stdenv.hostPlatform` I don't think is a major deparature from what you've got.

**Comment by aaronjanse** [comment]
> Ah, I didn't realize that. So it is TOML for sysroot, and JSON for everything else?

Correct

> You're doing great so far :), moving the config to stdenv.hostPlatform I don't think is a major deparature from what you've got.

Thanks! So I can figure out how to use this for Redox, how would I define a custom stdenv platform for the Redox kernel (which, unlike userspace, blacklists some instructions and lacks a libc)?

Redox userspace supports enough packages that I could create a PR to put that platform in Nixpkgs. The Redox kernel platform, however, would be used for nothing but the Redox kernel. Is there some way to define a stdenv.hostPlatform outside of Redox, then have that platform used for only one package (instead of needing to re-compile rustc etc)?

**Comment by Ericson2314** [comment]
Yeah! Do something like
```nix
import <nixpkgs> {
  crossSystem = {
    config = "x86_64-unknown-none-none"; # something like this
    platform.rustc = {
      name = "bare-metal-for-redox-or-whatever-you-want-to-call-it";
      ...jsonContents...
    };
  };
}
```
If the redox platform is stable, you can put it in `lib/systems/platforms.nix`. Basically that file is a grab-bag of stuff awaiting to be better organized, so I donn't really care if niche things are in there.

**Comment by Ericson2314** [comment]
C.f. https://github.com/mozilla/nixpkgs-mozilla/pull/222. Eventually, it would be nice if mozilla-nixpkgs could work well with this stuff to allow one to seamlessly switch between prebuilt sysroots (original made for rust-up) and nix-built ones.

**Comment by aaronjanse** [comment]
Oh awesome! Thanks for the explanation. I'll try to work on that this weekend :-)

File: pkgs/build-support/rust/default.nix
Comment (Mic92) [inline pkgs/build-support/rust/default.nix:78]: We have a rustSrc package in `nixpkgs/pkgs/development/compilers/rust/default.nix` already. Please re-use that.

File: pkgs/build-support/rust/default.nix
Comment (Mic92) [inline pkgs/build-support/rust/default.nix:102]: I would prefer if this parsing logic would be moved to the build time rather than evaluation time. Otherwise whenever nix wants to build this package it would also need to download the source for it, which is expensive on hydra and the final client that tries to download the binary package from the cache. Within nixpkgs we don't allow import-from-derivation for this reason.

File: pkgs/build-support/rust/default.nix
Comment (Mic92) [inline pkgs/build-support/rust/default.nix:96]: Could you move the whole expression to a different file?

File: pkgs/build-support/rust/default.nix
Comment (Mic92) [inline pkgs/build-support/rust/default.nix:126]: My recommendation here is to use some scripting language i.e. python with a toml library to build this to concatenate this.

File: pkgs/build-support/rust/default.nix
Comment (Mic92) [inline pkgs/build-support/rust/default.nix:129]: Is there a way to move `Cargo.toml` and `Cargo.lock` out of the nix file? That way there might also a chance to use `cargo` to update this lock file.

File: pkgs/build-support/rust/default.nix
Comment (Mic92) [inline pkgs/build-support/rust/default.nix:161]: I would say, move both `sysroot` and `sysrootSrc` to an external file.

File: pkgs/build-support/rust/default.nix
Comment (Mic92) [inline pkgs/build-support/rust/default.nix:37]: This needs documentation in `nixpkgs/doc/languages-frameworks/rust.section.md`
Also it should explain what the implication of custom sysroot is and when one 
wants to use it ideally with an example.

File: pkgs/build-support/rust/default.nix
Comment (Mic92) [inline pkgs/build-support/rust/default.nix:303]: So when do one use `targetIsJSON` but not `dontAddSysroot`?

**Comment by Mic92** [comment]
Can we have also an derivation that exercise this feature? Rust releases happens quite quickly. We currently have a testplan that includes building firefox/thunderbird on Linux and fd on macOS. We could bundle those up in a attrset and also include a test for custom target so that we can have one command that we use for testing every update.

**Comment by aaronjanse** [comment]
I feel like there should be some way to create a new pkgsCross without using `import`. Something like `pkgsCross.fromCrossSystem { ... }`

**Comment by aaronjanse** [comment]
I've figured out a way to handle Rust targets having information both general to a runtime environment (e.g. CPU architecture) and specific to a project (e.g. project-specific compile flags). I've also written code for generating most of the target information from other attributes in stdenv.hostPlatform.

I'm having trouble putting this idea into words at the moment, but I'm currently waiting on [1] to see if it will work for x86_64

[1] https://users.rust-lang.org/t/why-is-there-no-default-x86-64-unknown-none-target/47825

**Comment by aaronjanse** [comment]
I plan to push some fixes to this PR tomorrow or on Monday. Specifically, I plan to push code that uses Python to generate the sysroot Cargo.toml

File: pkgs/build-support/rust/sysroot/default.nix
Comment (Mic92) [inline pkgs/build-support/rust/sysroot/default.nix:16]: Cool. so this works now without `./tools`?

File: pkgs/build-support/rust/default.nix
Comment (Mic92) [inline pkgs/build-support/rust/default.nix:132]: I suppose these line are now obsolete?

File: pkgs/build-support/rust/sysroot/Cargo.lock
Comment (Mic92) [inline pkgs/build-support/rust/sysroot/Cargo.lock:8]: Do we need to take special care with updating this on rust releases?

**Comment by Ericson2314** [comment]
CC @lopsided98 you might be interested in this.

File: pkgs/build-support/rust/default.nix
Comment (aaronjanse) [reply pkgs/build-support/rust/default.nix:134]: Correct

File: pkgs/build-support/rust/sysroot/Cargo.lock
Comment (aaronjanse) [reply pkgs/build-support/rust/sysroot/Cargo.lock:8]: I'm not sure :-/

**Comment by Ericson2314** [comment]
@aaronjanse Could you add something in `pkgs/tests` so I can better play around with my diff from https://github.com/NixOS/nixpkgs/pull/95542#issuecomment-676850198 without breaking everything?

**Comment by aaronjanse** [comment]
I have midterms on Monday and Tuesday. I hope to be able to work on this again starting Wednesday. I hope to add the example OS described by https://os.phil-opp.com/ to the tests since following that tutorial is a very popular use of bare-metal Rust.

**Comment by aaronjanse** [comment]
Also, thank you @Ericson2314 for fixing the merge conflict!

**Comment by Ericson2314** [comment]
Good luck on the tests! We'll sync up after.

File: pkgs/build-support/rust/sysroot/Cargo.lock
Comment (aaronjanse) [reply pkgs/build-support/rust/sysroot/Cargo.lock:8]: Regardless, I'll add a script for easily updating the Cargo.lock

File: pkgs/build-support/rust/sysroot/default.nix
Comment (aaronjanse) [reply pkgs/build-support/rust/sysroot/default.nix]: Hmm, what's `./tools`?

Comment (aaronjanse) [reply pkgs/build-support/rust/sysroot/default.nix]: Oh, no, it still needs `./tools` in rustSrc. I've pushed code that adds it back again. I'll look into the size of `./tools` so we can decide if we want to add it to the existing nixpkgs rustSrc

**Comment by aaronjanse** [comment]
I just pushed complete (but not working) code for sysroot using Python. I'm having trouble debugging a cargoSha256 issue. I'll try to push a test this evening, but it might take a day or two to write.

@Ericson2314 I just realized that putting target.json info in stdenv.hostPlatform would work if I simply add more target.json defaults to embedded platforms than non-embedded platforms. Should I give that a shot?

**Comment by Ericson2314** [comment]
@aaronjanse Woo, more progress!

> @Ericson2314 I just realized that putting target.json info in stdenv.hostPlatform would work if I simply add more target.json defaults to embedded platforms than non-embedded platforms. Should I give that a shot?

Sure! That sounds good to me. But if you are stretched for time feel free to just focus on the main bits and I'll try to pitch in on that.

**Comment by aaronjanse** [comment]
> But if you are stretched for time feel free to just focus on the main bits and I'll try to pitch in on that.

Thank you @Ericson2314, that would be great. I'll continue working on getting a working test for this PR

**Comment by aaronjanse** [comment]
For now, a way to test this is by checking out [1], locally enabling nix-community cachix, then running `nix-build -A redox-vmdisk`.

[1] https://github.com/nix-community/redoxpkgs/tree/vmdisk/

**Comment by aaronjanse** [comment]
The most recent commit allows the sysroot to successfully compile. The problem was that a target may not support tests, which was fixed by adding `doCheck = false`. I also updated a cargoSha256 that was previously left broken.

Next up is making sure the sysroot works after being compiled then figuring out `pkgs/tests`.

**Comment by aaronjanse** [comment]
@Ericson2314 I'm still trying to figure out `pkgs/tests`, but here's my derivation for the Redox kernel. So far I've never had the kernel compile but not work during runtime, so for now, successfully building this derivation should be a strong predictor of whether sysroot code is correct.

```nix
{ rustPlatform, fetchFromGitHub, fetchgit }:

rustPlatform.buildRustPackage rec {
  name = "redox-kernel-latest";

  src = fetchgit {
    url = "https://gitlab.redox-os.org/redox-os/kernel";
    rev = "0590a71b8759c773cd961680ed9f4c35474bd748";
    sha256 = "1cdf5bc0b0fjr8k7p8maax8dm2w9i9jz5j64a579yxb4hy7zzhda";
    fetchSubmodules = true;
  };

  cargoSha256 = "19dim0py5sqg2a0ifjz1q9vzzfzlmw90n4wqlsc9yxc553kla728";

  target = src + /targets/x86_64-unknown-none.json;
  buildType = "debug";
  RUSTFLAGS = "-C debuginfo=2 -C soft-float -C lto=thin -C embed-bitcode=yes";
  RUSTC_BOOTSTRAP = 1;

  outputs = [ "out" "dev" ];
  dontPatchShebangs = true;
  dontStrip = true;

  doCheck = false;
}
```

Build with:
```bash
$ nix-build -E 'with import /path/to/this/pr/nixpkgs {}; callPackage ./default.nix {}'
```

File: pkgs/build-support/rust/sysroot/default.nix
Comment (Ericson2314) [inline pkgs/build-support/rust/sysroot/default.nix:5]: Did you try using `rustPlatform.rustcSrc` for this?

**Comment by Ericson2314** [comment]
OK! I added something along the lines of what I was thinking for managing the targets in Nix. With docs, too :).

BTW your redox example is missing `nasm`. Collapsed below is the fixed version
<details>
```nix
{ rustPlatform, fetchFromGitHub, fetchgit, nasm }:

rustPlatform.buildRustPackage rec {
  name = "redox-kernel-latest";

  src = fetchgit {
    url = "https://gitlab.redox-os.org/redox-os/kernel";
    rev = "0590a71b8759c773cd961680ed9f4c35474bd748";
    sha256 = "1cdf5bc0b0fjr8k7p8maax8dm2w9i9jz5j64a579yxb4hy7zzhda";
    fetchSubmodules = true;
  };

  outputs = [ "out" "dev" ];

  nativeBuildInputs = [ nasm ];

  cargoSha256 = "19dim0py5sqg2a0ifjz1q9vzzfzlmw90n4wqlsc9yxc553kla728";

  target = src + /targets/x86_64-unknown-none.json;
  buildType = "debug";
  RUSTFLAGS = "-C debuginfo=2 -C soft-float -C lto=thin -C embed-bitcode=yes";
  RUSTC_BOOTSTRAP = 1;

  dontPatchShebangs = true;
  dontStrip = true;

  doCheck = false;
}
```
</details>

What I wrote still needs tests as does everything else, but we can work on that next.

**Comment by Ericson2314** [comment]
@Kloenk and @lopsided98 it would be good to get your input on this one too.

**Comment by lopsided98** [comment]
I haven't given any review of this because I have no experience with embedded Rust. My use case for Rust cross-compilation is to target aarch32 NixOS machines. This looks fine to me, but I'm not very familiar with this area.

It does seem like future work could build off this to always build the sysroot as a separate derivation, allowing the same rustc to be used for native and cross-compiled builds.

**Comment by Kloenk** [comment]
Will try to test the following days

File: pkgs/build-support/rust/sysroot/default.nix
Comment (aaronjanse) [reply pkgs/build-support/rust/sysroot/default.nix:5]: Yes. The problem is that `rustPlatform.rustcSrc` deletes the `tools` directory (see below). Should I add a parameter (defaulting to current behavior) to `rustcSrc` to determine if the tools directory is deleted)

https://github.com/NixOS/nixpkgs/blob/359ac19b17085a6957b48d8e409ebc0e8f3b1daf/pkgs/development/compilers/rust/rust-src.nix#L9

File: pkgs/build-support/rust/sysroot/default.nix
Comment (aaronjanse) [reply pkgs/build-support/rust/sysroot/default.nix]: See https://github.com/NixOS/nixpkgs/pull/95542#discussion_r504500893

File: pkgs/build-support/rust/default.nix
Comment (aaronjanse) [reply pkgs/build-support/rust/default.nix]: Oh wow, only 17 MB

**Comment by aaronjanse** [comment]
Thank you for the commits @Ericson2314!

> We should write that small blog os test soon too.

Definitely. Half of my time following the blog os tutorial was getting sysroot stuff to work (which is by no means the fault of the tutorial author who kindly spends a lot of time helping readers debug issues).

I just pushed code that keeps `tools` and `stdarch` in rustcSrc then uses rustcSrc for the sysroot, increasing the size of the rustcSrc derivation output from 13 MB to 32 MB. Is this acceptable?

**Comment by aaronjanse** [comment]
Oh, and Redoxpkgs status update for those interested: I recently got a derivation working to generate a Redox VM with Nixpkgs software pre-installed and callable from the $PATH. Next up after this PR is packaging Redox graphics stuff and getting Nix itself to run on Redox.

**Review by kloenk — APPROVED** [review]
This looks good but did not yet get around to test it.

**Comment by Kloenk** [comment]
I think this should target staging, as it's triggering a mass rebuild

**Comment by Kloenk** [comment]
Is there a way to get the source in the sysroot? This would be needed for [std aware cargo](https://github.com/rust-lang/wg-cargo-std-aware) and I did not see a flag to enable it.

**Comment by Kloenk** [comment]
@Ericson2314 not sure. As far as I understood, it has to have a special tree and not just the rustc source. Also cargo still seems to be missing features to set where to search for the source. I think cargo wants it in $(which rustc)/../lib or so.

**Comment by Ericson2314** [comment]
@Kloenk well the standard library source is in that repo. Perhaps Cargo is making some stupid assumptions we need to patch, or if it was transformed a bit for the rustup version we need to replicate that as transformation as well.

**Comment by Kloenk** [comment]
> Note that for this to work you need to have the source code for the standard library available, and at this time the only supported method of doing so is to add the rust-src rust rustup component:

taken from https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#build-std

**Comment by Ericson2314** [comment]
Well, a dearth of "supported methods" shouldn't stop us :D

**Comment by aaronjanse** [comment]
> Maybe just put the new behavior a flag for now and override just for your case so we can send this to master?

Done

File: pkgs/build-support/rust/default.nix
Comment (Ericson2314) [inline pkgs/build-support/rust/default.nix:205]: @aaronjanse I think this indentation here might also be causing the rebuild?

Comment (aaronjanse) [reply pkgs/build-support/rust/default.nix]: Fixed!

**Comment by Ericson2314** [comment]
Excellent! I'm tempted to merge now, but we should probably have that test first?

**Comment by aaronjanse** [comment]
I'm busy for the next few days, but next week I'll have the time & motivation to write a blog post about how to follow https://os.phil-opp.com/ using Nix.

I'd be happy to write up a full qemu blog os test some time next week. Maybe we could do a simple "Rust compiles blog os kernel (or redox kernel, for now) with exit code zero" for now then a follow-up PR next week that'll include a test with qemu to demo usage of this feature (and link to the test in Nixpkgs's Rust docs)?

File: pkgs/build-support/rust/sysroot/Cargo.lock
Comment (Ericson2314) [reply pkgs/build-support/rust/sysroot/Cargo.lock:8]: Do you have a plan for this? Writing that script sounds non-trivial to me!

**Comment by Ericson2314** [comment]
@aaronjanse That all sounds amazing, but to be fair is well beyond what merging this is blocked on. A simple "does it build" test, no qemu necessary, would suffice for now.

Of course it's great to do the qemu part too! (And I think @matthewbauer wrote some prior qemu tests that might be useful to ok at.) But I'm fine merging that in a follow-up PR.

**Comment by aaronjanse** [comment]
Sweet, in that case I'll work on a test for this PR some time late tonight (or early tomorrow, based on time zone) :-)

File: pkgs/build-support/rust/sysroot/Cargo.lock
Comment (aaronjanse) [reply pkgs/build-support/rust/sysroot/Cargo.lock:8]: I don't have a plan for updating the version numbers in Cargo.toml, but I plan to at least write an update script to generate the Cargo.lock from the Python script (using Cargo)

Speaking of which, is there some place I should put my name so I'm notified when there are `build-support/rust` PRs in Nixpkgs? That would make it easier for me to maintain the sysroot code :-)

**Comment by aaronjanse** [comment]
I've added a test for compiling blog_os and cleaned up a little. I put these in changes small commits so that they're easy to review.

I added `assert useSysroot -> !(args.doCheck or true)` and relevant documentation stating that tests won't work because custom sysroots are being built without the `std` crate (is this a blocking issue for this PR?).

I also pushed a commit (https://github.com/NixOS/nixpkgs/pull/95542/commits/d884b2d877cfda8247a6ee49f5c26ba3133c3889) that sets the `RUSTFLAGS` environment variable as a derivation attribute instead of injecting it into buildPhase. This seems cleaner, but I'd like for the change to be reviewed since it's a notable last-minute change. I'd be happy to revert this commit if so desired.

P.S. Feel free to rename this PR and the final commit to Nixpkgs to `buildRustPackage: support custom targets`. I forgot where the original `makeRustPlatform` text in the title came from.

**Comment by aaronjanse** [comment]
Note that the new `update-lockfile.sh` script doesn't update the `cargoSha256`.

File: doc/languages-frameworks/rust.section.md
Comment (aaronjanse) [inline doc/languages-frameworks/rust.section.md:127]: Is having the user explicitly disable tests better than silently disabling them?

Comment (Kloenk) [reply doc/languages-frameworks/rust.section.md:127]: Just to make sure, which tests are you talking about? Testing if rustc works, or the rust test crate?

Comment (aaronjanse) [reply doc/languages-frameworks/rust.section.md:127]: This refers to the checkPhase of buildRustPackage, where `cargo test` fails to run on users' code:

```
λ nix-shell . -A rustPackageWithCustomSysroot
$ cargo build # success
$ cargo test  # error
```

File: pkgs/build-support/rust/sysroot/Cargo.lock
Comment (aaronjanse) [reply pkgs/build-support/rust/sysroot/Cargo.lock]: > Speaking of which, is there some place I should put my name so I'm notified when there are build-support/rust PRs in Nixpkgs?

Ah, looks like that place is codeowners, but GitHub requires excessive permissions for that to be useful. I plan to follow nixpkgs-unstable while locally working on Redox, so I should catch some sysroot regressions when they happen. More sysroot testing in the future could also help prevent issues.

**Comment by Kloenk** [comment]
FYI: I have the feeling rust did some breaking changes to sysroot. Not sure if this will affect this pr (some folders are renamed)

**Comment by Ericson2314** [comment]
ugh I missed the test!

**Review by Ericson2314 — APPROVED** [review]
Finally!

**Pull request merged**
